Scripts/Editor/Debug/ScriptDefinesToggle.cs

```diff
@@ -0,0 +1,99 @@
+using System.Collections.Generic;
+using System.Linq;
+using UnityEditor;
+using UnityEditor.Build;
+
+namespace Anvil.Unity.Editor.Debug
+{
+    /// <summary>
+    /// Editor helper class to easily toggle on and off Script Defines and their dependencies.
+    /// </summary>
+    [InitializeOnLoad]
+    public static class ScriptDefinesToggle
+    {
+        /// <summary>
+        /// The base location for where items should appear in the menu.
+        /// </summary>
+        public const string MENU_PATH_BASE = "Anvil/Script Defines";
+        
+        private static readonly NamedBuildTarget NAMED_BUILD_TARGET;
+        private static readonly Dictionary<string, ScriptDefineDefinition> SCRIPT_DEFINE_DEFINITIONS_LOOKUP;
+
+        static ScriptDefinesToggle()
+        {
+            SCRIPT_DEFINE_DEFINITIONS_LOOKUP = new Dictionary<string, ScriptDefineDefinition>();
+            
+            BuildTarget buildTarget = EditorUserBuildSettings.activeBuildTarget;
+            BuildTargetGroup buildTargetGroup = BuildPipeline.GetBuildTargetGroup(buildTarget);
+            NAMED_BUILD_TARGET = NamedBuildTarget.FromBuildTargetGroup(buildTargetGroup);
+        }
+        
+        /// <summary>
+        /// Validator function to show if a given Script Define is currently enabled or not.
+        /// Call this to validate a Script Define menu item.
+        /// </summary>
+        /// <param name="scriptDefineDefinition">The <see cref="ScriptDefineDefinition"/> representing the define</param>
+        /// <returns>True as Script Defines can be toggled at any time</returns>
+        public static bool Toggle_Validator(ScriptDefineDefinition scriptDefineDefinition)
+        {
+            PlayerSettings.GetScriptingDefineSymbols(NAMED_BUILD_TARGET, out string[] defines);
+            bool isEnabled = defines.Contains(scriptDefineDefinition.Define);
+            Menu.SetChecked(scriptDefineDefinition.MenuPath, isEnabled);
+            return true;
+        }
+        
+        /// <summary>
+        /// Toggle function to turn on or off a given Script Define.
+        /// Call this to toggle a Script Define from the menu.
+        /// </summary>
+        /// <param name="scriptDefineDefinition">The <see cref="ScriptDefineDefinition"/> representing the define</param>
+        public static void Toggle(ScriptDefineDefinition scriptDefineDefinition)
```

Comment:
In a simple case I think this works fine but what if two different `ScriptDefineDefinition` instances have the same define as a requirement?

If both are enabled and you disable one the required defines will be removed for the one define that is still enabled.

As we make use of this I see a lot of options clobbering each other's defines and having a hard time tracking it down.

I don't know if there's a way to do this nicely without getting every define to buy into a system. We may be better off just asking developers to enable and dependent conditions with a `#if MY_HIGH_LEVEL_TOGGLE && !MY_DEPENDENCY` condition somewhere in the code that reports the requirement. Enable isn't a problem but cleaning up/disabling dependencies is going to be a manual process I think.

Luckily a git diff will generally catch it easily.

Reply:
Ok this was a good point and made me hem and haw a bit over just getting something simple in vs making a fully blown Editor panel that allows for a ton of configuration and rules and error checking.

I built a test case and it ended up being pretty quick to make something more robust but still just a menu item. Full blown panel can come later if we decide to get indulgent. 

I have a define called `BASE` which for example purposes is not owned by us. 

Then a define called `FEATURE_A` which is owned by us and requires `BASE` to be set.
Then a define called `FEATURE_B` which is also owned by us and requires `BASE` to be set. 

Both `FEATURE_A` and `FEATURE_B` appear in the menu because we defined them and said we want them to be there. 

Finally there is another define called `DERIVED` which is not owned by us but both `FEATURE_A` and `FEATURE_B` have said that it is dependent on them. 

In a blank slate, nothing is on.

Go to the menu, toggle `FEATURE_A`. 
The script turns on `FEATURE_A` and also turns on `BASE` because it's a requirement. 
Go to the menu, toggle `FEATURE_B`.
The script turns on `FEATURE_B` and `BASE` is already on. 

Now go into the project settings and manually remove `BASE`. This simulates someone manually messing with things, another script turning it off, or another third party library with their own script define toggles turning off something we rely on. 

The script detects the script define change, realizes that we have `FEATURE_A` and `FEATURE_B` that depended on `BASE` which is now gone and turns them off too. 

On the `DERIVED` side, i think it's a bit of stretch edge case. Why would a third party define require one or more defines in our framework? But still.

Manually go and add `DERIVED` to project settings. Simulating again, someone messing with things, other scripts or frameworks. 

`DERIVED` requires `FEATURE_A` and `FEATURE_B` to be set which both require `BASE` to be set. They aren't, so we get a helpful error in the console telling us to enable those requirements. `DERIVED` is not set.

Finally, assume everyone is set. Then manually remove `BASE`. All of `BASE`s dependent defines are now removed as well. 

The only thing we don't handle is exclusions. Such as if `FEATURE_A` is enabled, `FEATURE_B` must NOT be enabled. Maybe we could add support for that in the future but I feel like the developer of `FEATURE_A` is going to build in some sort of guard in their own code anyway to yell at the developer if both are enabled anyway. Seems like an unlikely scenario given how Script Defines in C# work anyway.

Clobbering should now not be an issue. Everything, (our own defines and existing or third party defines) have requirements and dependencies merged and the toggling logic is explicit when you are making an explicit choice through the menu:

- Enable all requirements when enabling
- Disable all dependants when disabling

But it's lenient when an implicit choice happens through manually editing or a different third party script changes.

- Log to console necessary requirements when enabling
- Disable all dependants when disabling

This leaves you always in a "Will compile" state.

Reply:
> The only thing we don't handle is exclusions. Such as if `FEATURE_A` is enabled, `FEATURE_B` must NOT be enabled. Maybe we could add support for that in the future but I feel like the developer of `FEATURE_A` is going to build in some sort of guard in their own code anyway to yell at the developer if both are enabled anyway. Seems like an unlikely scenario given how Script Defines in C# work anyway.

That's fine. I agree we don't need this feature.

> - Disable all dependants when disabling

In your example if `FEATURE_A` and `FEATURE_B` are enabled then I disable `FEATURE_B`. `BASE` will remain right? Because `FEATURE_A` still needs it.

Reply:
Correct!